Add 'gdal raster update' (#12378)
OSGeo/gdal: doc/source/programs/gdal_raster_update.rst

.. _gdal_raster_update:

================================================================================
``gdal raster update``
================================================================================

.. versionadded:: 3.12

.. only:: html

    Update the destination raster with the content of the input one.

.. Index:: gdal raster update

Synopsis
--------

.. program-output:: gdal raster update --help-doc

Description
-----------

:program:`gdal raster update` can be used to update an existing target raster dataset
with the pixel values of the source raster, doing reprojection if the source
and target datasets do not have the same coordinate reference systems.
The extent, size, resolution or coordinate reference system of the target dataset
are not modified by this operation.

> Overviews are *not* updated by this command, but may be if using :ref:`gdal_raster_overview_refresh`.

Reviewer: Any thoughts on inverting this behavior? Update any existing overviews unless `--no-update-overviews` is specified?

Author: > Update any existing overviews unless `--no-update-overviews` is specified?

done

Standard options
++++++++++++++++

> .. option:: -r, --resampling <RESAMPLING>

Reviewer: Worth pulling this out into a separate file that can be included?

Author: done

    ``sum``: compute the weighted sum of all non-NODATA contributing pixels

    .. note::

        When downsampling is performed (use of :option:`--resolution` or :option:`--size`), existing
        overviews (either internal/implicit or external ones) on the source image
        will be used by default by selecting the closest overview to the desired output
        resolution.
        The resampling method used to create those overviews is generally not the one you
        specify through the :option:`-r` option.

.. option:: --geometry <WKT_or_GeoJSON>

    Geometry as a WKT or GeoJSON string of a polygon (or multipolygon) to which
    to restrict the update.
    If the input geometry is GeoJSON, its CRS is assumed to be WGS84, unless there is
    a CRS defined in the GeoJSON geometry or :option:`--geometry-crs` is specified.
    If the input geometry is WKT, its CRS is assumed to be the one of the input dataset,
    unless :option:`--geometry-crs` is specified.
    The X and Y axis are the "GIS friendly ones", that is X is longitude or easting,
    and Y is latitude or northing.

.. option:: --geometry-crs <CRS>

>     CRS in which the coordinates values of :option:`--geometry`

Reviewer: ```suggestion
    CRS in which the coordinate values of :option:`--geometry`
```

Author: fixd

    are expressed. If not specified, it is assumed to be the CRS of the input
    dataset.

>     The bounds are reprojected from the geometry-crs to the CRS of the
>     input dataset.

Reviewer: ```suggestion
    The geometry will be reprojected from the geometry-crs to the CRS of the
    dataset being updated.
```

?

Author: fixed

Advanced options
++++++++++++++++

.. option:: --wo, --warp-option <NAME>=<VALUE>

    Set a warp option.  The :cpp:member:`GDALWarpOptions::papszWarpOptions` docs show all options.
    Multiple options may be listed.

> .. option:: -to <NAME>=<VALUE>

Reviewer 2: Should it be `--to`?

    Set a transformer option suitable to pass to :cpp:func:`GDALCreateGenImgProjTransformer2`.
    See :cpp:func:`GDALCreateRPCTransformerV2()` for RPC specific options.

> .. option:: --et, --error-threshold <ERROR-THRESHOLD>

Reviewer: Beyond the scope of this PR, but I'm wondering it's worth reconsidering the default value for `--error-threshold`. We haven't found a huge performance boost from the default value, and have had some surprising errors introduced with non-zero values.

Author: > and have had some surprising errors introduced with non-zero values.

I'd be curious about the exact situation where that would happen

Reviewer: I'll see what I can dig up. Maybe the issue is the assumption that the error is greatest at the midpoint between the points that are actually transformed. Would it make sense that that is not always true for certain transformations?

Author: > Would it make sense that that is not always true for certain transformations?

I can imagine that could happen for polar projections <--> geographic

Author: > Should GDAL try to be smart and disable the approximator for such cases?

We have already some exceptions/tuning for such cases in various places (look for "pole" in gdaltransformer.cpp, gdalwarper.cpp, gdalwarpoperation.cpp). It is hard to do in the general case however
The approximate transformer helps substantially for UTM <--> geographic since the accurate Transverse Mercator we use is not cheap, especially if using nearest neighbour where the limiting factor becomes PROJ
